tests/src/integration/client.rs

```diff
@@ -0,0 +1,664 @@
+                        )
+                        .await;
+                    }
+                    cmds::ValidatorSet::ValidatorSetUpdateRelay(args) => {
+                        let eth_client = Arc::new(
+                            Provider::<Http>::try_from(&args.eth_rpc_endpoint)
+                                .unwrap(),
+                        );
+                        let args = args.to_sdk_ctxless();
+                        validator_set::relay_validator_set_update(
+                            eth_client, self, args,
+                        )
+                        .await
+                        .proceed_or_else(|| eyre!("Fatal error"))?;
+                    }
+                },
+            },
+        }
+        Ok(())
+    }
+}
+
+struct TempFile(PathBuf);
+impl TempFile {
+    fn new(path: PathBuf) -> (Self, File) {
+        let f = File::create(&path).unwrap();
+        (Self(path), f)
+    }
+}
+
+impl Drop for TempFile {
+    fn drop(&mut self) {
+        _ = std::fs::remove_file(&self.0);
+    }
+}
+
+/// Test helper that captures stdout of
+/// a process.
+pub struct CapturedOutput<T = ()> {
+    pub output: String,
+    pub result: T,
+    input: String,
+}
+
+impl CapturedOutput {
+    pub fn with_input(input: String) -> Self {
+        Self {
+            output: "".to_string(),
+            result: (),
+            input,
+        }
+    }
+}
+
+impl<T> CapturedOutput<T> {
+    pub(crate) fn of<F>(func: F) -> Self
+    where
+        F: FnOnce() -> T,
+    {
+        std::io::set_output_capture(Some(Default::default()));
+        let mut capture = Self {
+            output: Default::default(),
+            result: func(),
+            input: Default::default(),
+        };
+        let captured = std::io::set_output_capture(None);
+        let captured = captured.unwrap();
+        let captured = Arc::try_unwrap(captured).unwrap();
+        let captured = captured.into_inner().unwrap();
+        capture.output = String::from_utf8(captured).unwrap();
+        capture
+    }
+
+    pub fn run<U, F>(&self, func: F) -> CapturedOutput<U>
+    where
+        F: FnOnce() -> U,
+    {
+        use std::io::Write;
+        let _temp = {
+            let (temp, mut f) = TempFile::new(PathBuf::from("stdin.mock"));
+            write!(&mut f, "{}", self.input).unwrap();
+            temp
+        };
+        CapturedOutput::of(func)
+    }
+
+    /// Check if the captured output contains the regex.
+    pub fn matches(&self, needle: regex::Regex) -> bool {
+        needle.captures(&self.output).is_some()
+    }
+
+    /// Check if the captured output contains the string.
+    pub fn contains(&self, needle: &str) -> bool {
+        let needle = regex::Regex::new(needle).unwrap();
+        self.matches(needle)
+    }
+}
+
+async fn submit_transfer(
+    client: &MockNode,
+    ctx: &mut Context,
+    args: args::TxTransfer,
+) -> Result<(), namada::ledger::tx::Error> {
+    let arg = args.clone();
```

Comment:
also this would be better to run through the original without a mock version

Reply:
Yeah, I wanted to, but that has problems. It submits txs twice in case masp hits an epoch boundary condition. This greatly complicates checking if the tx succeeded.

Reply:
But I'm open to suggestions!

Reply:
Actually, I think I can use the original if I'm super careful about epoch boundaries in the tests.